---
title: "Diversity of Bacterial Communities in Wetlands of Calakmul Biosphere Reserve: A Comparative Analysis between Conserved and Semi-Urbanized Zones in Pre-Mayan Train Era"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Diversity of Bacterial Communities in Wetlands of Calakmul Biosphere Reserve: A Comparative Analysis between Conserved and Semi-Urbanized Zones in Pre-Mayan Train Era

**Authors:**
1. David A. García-Estrada, Unidad de Genómica Avanzada, Centro de Investigación y Estudios Avanzados del Instituto Politécnico Nacional, Irapuato, Guanajuato, México.
2. Nelly Selem-Mojica, Centro de Ciencias Matemáticas, Universidad Nacional Autónoma de México (UNAM), Morelia, Mich. México
3. Aída Martínez-Hernández, Colegio de Postgraduados Campus Campeche, Sihochac, Champotón, Camp. México 
4. Joel Lara-Reyna, Colegio de Postgraduados Campus Campeche, Sihochac, Champotón, Camp. México 
5. Sonia Dávila-Ramos, Centro de Investigación en Dinámica Celular, IICBA, Universidad Autónoma del Estado de Morelos (UAEM), Cuernavaca, Mor. México
6. Karina Verdel-Aranda, CONACyT-Colegio de Postgraduados Campus Campeche, Sihochac, Champotón, Camp. México
7. Present address: Tecnológico Nacional de México-Instituto Tecnológico de Chiná, Chiná, Camp. México.

**Corresponding author:** Karina Verdel-Aranda, e-mail: karina.va@china.tecnm.mx

## Abstract

**Background**: The Calakmul Biosphere Reserve (CBR) is known for its rich animal and plant biodiversity, yet its microbial communities remain largely unknown. The reserve does not possess permanent bodies of water, nevertheless, seasonal depressions associated with fractures create wetlands, known locally as aguadas. Given the recent construction of the Maya train that crosses the CRB, it is essential to assess the biodiversity of its microorganisms and recognize their potential as a valuable source of goods. This evaluation is pivotal in mitigating potential mismanagement of the forest ecosystem. To enhance comprehension of microbial communities, we characterized the microbiota in three different wetlands. Ag-UD1 and Ag-UD2 wetlands are located in a zone without human disturbances, while the third, Ag-SU3, is situated in a semi-urbanized zone. Sampling was carried out over a three-year period (2017, 2018, and 2019), enabling the monitoring of spatiotemporal variations in bacterial community diversity. The characterization of microbiome composition was conducted using 16S rRNA metabarcoding. Concurrently, the genomic potential of select samples was examined through shotgun metagenomics. 
**Results**: Statistical analysis of alpha and beta diversity showed significant differences among the bacterial communities found in undisturbed sites Ag-UD1 and Ag-UD2, in comparison to Ag-SU3. However, no significant differences were observed among sites belonging to the undisturbed area. Furthermore, a comparative analysis at the zone level reveals substantial divergence among the communities, indicating that the geographic location of the samples significantly influences these patterns. The bacterial communities in the CBR wetlands predominantly consist of genera from phyla Actinobacteria, Acidobacteria and Proteobacteria.
**Conclusion**: This characterization has identified the composition of microbial communities and provided the initial overview of the metabolic capacities of the microbiomes inhabiting the aguadas across diverse conservation zones. The three sites exhibit distinct microbial compositions, suggesting that variables such as chemical composition, natural and anthropogenic disturbances, vegetation, and fauna may play a pivotal role in determining the microbial structure of the aguadas. This study establishes a foundational baseline for evaluating the impact of climatic factors and human interventions on critical environments such as wetlands.

# Script rRNA16s
## Metadata and Raw data processing

**Order of readings**

```{r}
# Sort file names
fnFs <- sort(list.files(RawData, pattern = "_R1.fastq.gz"))
fnRs <- sort(list.files(RawData, pattern = "_R2.fastq.gz"))
```

**Filtering and cutting**

```{r}
filtFs <- file.path(filt_path, paste0(sampleNames, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sampleNames, "_R_filt.fastq.gz"))

# Filter and Trimming
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen = c(250, 230),
              maxN = 0, maxEE = c(2, 2), truncQ = 2, rm.phix = TRUE,
              compress = TRUE, multithread = FALSE)
```

**De-replicate**

```{r message=FALSE, warning=FALSE}
# De-replicate
derepFs <- derepFastq(filtFs, verbose = TRUE)
derepRs <- derepFastq(filtRs, verbose = TRUE)

# Add the names of the samples to the de-replicated object
names(derepFs) <- sampleNames
names(derepRs) <- sampleNames
```

Proceed to generate an error model.

```{r}
errF <- learnErrors(filtFs, multithread = FALSE)
errR <- learnErrors(filtRs, multithread = FALSE)
```

**Amplicon Sequence Variants (ASVs)**

```{r}
dadaFs <- dada(derepFs, err = errF, multithread = FALSE, pool = TRUE)
dadaRs <- dada(derepRs, err = errR, multithread = FALSE, pool = TRUE)
```

**Merge paired reads**

```{r}
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs)
```

**Build sequence table**

```{r}
seqtabAll <- makeSequenceTable(mergers[!grepl("Mock", names(mergers))])
dim(seqtabAll)
table(nchar(getSequences(seqtabAll)))
```

**Remove chimeras**

```{r}
seqtabNoC <- removeBimeraDenovo(seqtabAll, method = "consensus",
                                multithread = FALSE, verbose = TRUE)
dim(seqtabNoC)
sum(seqtabNoC)/sum(seqtabAll)
```

**Metadata**

```{r}
samdf <- read.csv("Metadata16s.csv", header = TRUE, row.names = 1)
all(rownames(seqtabAll) %in% rownames(samdf))
```

## Taxonomic Assignment and Phylogenetic tree

**SILVA version 138**

```{r}
fastaRef <- "silva_nr_v138_train_set.fa"
taxTab <- assignTaxonomy(seqtabNoC, refFasta = fastaRef, multithread = TRUE)
taxTabExtra <- addSpecies(taxTab, "silva_species_assignment_v138.fa",
                          verbose = TRUE)
unname(head(taxTab)) -> tabla
colnames(tabla) <- c("Kingdom", "Phylum", "Order", "Class", "Family", "Genus")
```

**Phylogenetic tree**

```{r}
seqs <- getSequences(seqtabNoC)
names(seqs) <- seqs
alignment <- AlignSeqs(DNAStringSet(seqs), anchor = NA, verbose = FALSE)
phangAlign <- phyDat(as(alignment, "matrix"), type = "DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) 
fit <- pml(treeNJ, data = phangAlign)
fitGTR <- update(fit, k = 4, inv = 0.2)
```

## Data transformation and manipulation

**Prune and subset taxa**

```{r, message = FALSE, warning = FALSE}
summary(psd5@tax_table@.Data == "")
psd5 <- prune_taxa(taxa_sums(psd5) > 0, psd5)
Phylo_P <- subset_taxa(psd5, Phylum != "")
Phylo_G <- subset_taxa(psd5, Genus != "")
summary(Phylo_P@tax_table@.Data == "")  
summary(Phylo_G@tax_table@.Data == "")  
```

**Rarefacction data**

```{r message=FALSE, warning=FALSE}
# Grouping by taxonomic level
Phylum_A <- tax_glom(physeq = Phylo_P, taxrank = "Phylum", NArm = F)
Genus_A <- tax_glom(physeq = Phylo_G, taxrank = "Genus", NArm = F)
# Rarefaction Species Richness
Phylum_RR <- rarefy_even_depth(Phylum_A, sample.size = min(sample_sums(Phylum_A)),
                               rngseed = 711, replace = TRUE, 
                               trimOTUs = TRUE, verbose = TRUE)
Genus_RR <- rarefy_even_depth(Genus_A, sample.size = min(sample_sums(Genus_A)),
                              rngseed = 711, replace = TRUE, 
                              trimOTUs = TRUE, verbose = TRUE)
# Rarefaction relative data
Phylum_RRR <- transform_sample_counts(Phylum_RR, function(x) round(x*100 / sum(x), 0))
Genus_RRR <- transform_sample_counts(Genus_RR, function(x) round(x*100 / sum(x), 0))
```

**Rarefaction Relative data**

```{r, message = FALSE, warning = FALSE}
Phylum_RRR_DF <- psmelt(Phylum_RRR)
Genus_RRR_DF <- psmelt(Genus_RRR)
# Grouping 
Phylum_RRR_DF$Phylum[Phylum_RRR_DF$Abundance <= 1] <- "Phylum <= 1 % abundance"
Genus_RRR_DF$Phylum[Genus_RRR_DF$Abundance <= 1] <- "Phylum <= 1 % abundance"
Genus_RRR_DF$Genus[Genus_RRR_DF$Abundance <= 2] <- "Genus <= 2 % abundance"
```

## Abundance Plot

```{r}
# Plot Phylum
tx <- "Phylum"
plot <- ggplot(data = Phylum_RRR_DF, aes(x = SITE, y = Abundance, fill = get(tx))) +
  geom_bar(stat = "identity", position = "fill") + 
  # Scale
  labs(title = paste("Relative abundance (Rarefaction) at the taxonomic level of", tx, 
                     "in the Phyloseq object agglomerate taxa at the Phylum level"),
       x = "Samples", y = "Relative abundance (Rarefaction)", fill = tx) +
  facet_grid(. ~ Phylum_RRR_DF$YEAR)
print(plot)

# Plot Genus
leveltax <- c("Phylum", "Genus")
for (tx in leveltax){
  # tx <- "Phylum"
  plot <- ggplot(data = Genus_RRR_DF, aes(x = SITE, y = Abundance, fill = get(tx))) +
    geom_bar(stat = "identity", position = "fill") + 
    # Scale
    labs(title = paste("Relative abundance (Rarefaction) at the taxonomic level of", tx, 
                       "in the Phyloseq object agglomerate taxa at the Genus level"),
         x = "Samples", y = "Relative abundance (Rarefaction)", fill = tx) +
    facet_grid(. ~ Genus_RRR_DF$YEAR)
  print(plot)
}
```

## Diversity analysis

**Alpha diversity**

```{r}
obj.phylo <- c("Phylum_RR", "Genus_RR")
measures <- c("Observed", "Chao1", "Shannon", "Simpson")
for (o.p in obj.phylo){
  # o.p <- "Genus_A"
  p.alpha <- plot_richness(physeq = get(o.p), color = "SITE", x = "SITE", 
                           measures = measures,
                           title = paste("Alpha diversity indices for the Calakmul",
                                          "samples obtained by rRNA 16s of", o.p, "data")) +
    labs(x = "Site", y = "Alpha Diversity Measure")
  print(p.alpha)
  dev.off()
}
```

**Beta Diversity**

```{r}
o <- "PCoA"
d <- "wunifrac"
for (o.p in obj.phylo){
  Ord <- ordinate(physeq = get(o.p), method = o, distance = d)
  Beta <- plot_ordination(physeq = get(o.p), ordination = Ord, 
                          color = "SITE_M", shape = "YEAR") +
    geom_point(size = 5) +
    geom_text(mapping = aes(label = SITE), size = 4, vjust = 2, hjust = 1) + 
    geom_vline(xintercept = 0) +
    geom_hline(yintercept = 0) +
    labs(title = paste("Beta diversity with", o, "-", d, 
                       "for the Calakmul samples rRNA 16s", o.p))
  plot(Beta)
}
```

## Hypothesis test

**Diversity alpha**

```{r}
measures <- c("Observed", "Chao1", "Shannon", "Simpson")
sta.results <- data.frame(Obj.Phyloseq = character(),
                          Measure = character(),
                          Site1 = character(),
                          Site2 = character(),
                          p_value = numeric(), 
                          adjust_method = character())
sta.results.tukey <- data.frame(Obj.Phyloseq = character(),
                                Measure = character(),
                                Versus = character(),
                                p_value = numeric())
for (o.p in obj.phylo){
  tryCatch({
    # o.p <- "Genus_R"
    ob.phy <- get(o.p)
    OTU <- ob.phy@otu_table@.Data
    SAM <- ob.phy@sam_data
    SAM$Sample <- rownames(SAM)
    ## Observed
    Obs <- estimateR(OTU)[1,]
    ObsDF <- data.frame(Sample = names(Obs), Observed = Obs)
    ## Chao1
    Chao <- estimateR(OTU)[2,]
    ChaoDF <- data.frame(Sample = names(Chao), Chao1 = Chao)
    ## Shannon
    Shannon <- vegan::diversity(OTU, "shannon")
    ShannonDF <- data.frame(Sample = names(Shannon), Shannon = Shannon)
    ## Simpson
    Simpson <-  vegan::diversity(OTU, "simpson")
    SimpsonDF <- data.frame(Sample = names(Simpson), Simpson = Simpson)
    # Join Diversity with Metadata
    Diversity <- inner_join(ObsDF, ChaoDF, by = "Sample") %>% 
      inner_join(., ShannonDF, by = "Sample") %>% 
      inner_join(., SimpsonDF, by = "Sample") %>%
      inner_join(., SAM, by = "Sample")
    Diversity2 <- Diversity %>% 
      select(SITE, Observed, Chao1, Shannon, Simpson) %>%
      gather("Index", "Value", -SITE)
    # Mean Diversity 
    MeanDiversity <- ddply(Diversity, "SITE", summarise, 
                           MeanObserved = mean(Observed), 
                           MeanChao1 = mean(Chao1),
                           MeanShannon = mean(Shannon),
                           MeanSimpson = mean(Simpson)
                           )
    MeanLines <- ddply(Diversity2, c("SITE", "Index"), summarise, 
                   MeanValue = mean(Value))
    for (idx in measures){
      # idx <- "Shannon"
      # Fitting linear models
      fit <- lm(formula = Diversity[[idx]] ~ Diversity[["SITE"]], data = Diversity)
      # ANOVA
      ANOVA <- anova(fit)
      ANOVA
      # Intercepto
      aov_fit <- aov(formula = Diversity[[idx]] ~ Diversity[["SITE"]], data = Diversity)
      summary(aov_fit, intercept = T)
      # pairwise.t.test - no adjust
      t.test.NA <- pairwise.t.test(Diversity[[idx]], Diversity$SITE, 
                                   p.adjust = "none", pool.sd = T)
      t.test.NA.melt <- melt(t.test.NA$p.value, na.rm = TRUE)
      colnames(t.test.NA.melt) <- c("Site1", "Site2", "p_value")
      sta.results <- bind_rows(sta.results, 
                               data.frame(Obj.Phyloseq = o.p,
                                          Measure = idx, 
                                          Site1 = t.test.NA.melt$Site1,
                                          Site2 = t.test.NA.melt$Site2,
                                          p_value = t.test.NA.melt$p_value,
                                          adjust_method = "None"))
      # pairwise.t.test - Bonferroni adjust
      t.test.Bon <- pairwise.t.test(Diversity[[idx]], Diversity$SITE, 
                                    p.adjust = "bonferroni", pool.sd = T)
      t.test.Bon.melt <- melt(t.test.Bon$p.value, na.rm = TRUE)
      colnames(t.test.Bon.melt) <- c("Site1", "Site2", "p_value")
      sta.results <- bind_rows(sta.results, 
                               data.frame(Obj.Phyloseq = o.p,
                                          Measure = idx, 
                                          Site1 = t.test.Bon.melt$Site1,
                                          Site2 = t.test.Bon.melt$Site2,
                                          p_value = t.test.Bon.melt$p_value,
                                          adjust_method = "Bonferroni"))
      # Tukey
      tukey <- TukeyHSD(aov_fit, conf.level = 0.95)
      sta.results.tukey <- bind_rows(sta.results.tukey, 
                                     data.frame(Obj.Phyloseq = o.p,
                                                Measure = idx,
                                                Versus = names(tukey$`Diversity[["SITE"]]`[,4]),
                                                p_value = tukey$`Diversity[["SITE"]]`[,4]))
    }
  }, error = function(e){cat("ERROR :",conditionMessage(e),"\n")})
}
# Significative results
sta.results$Versus <- paste(sta.results$Site1, sta.results$Site2, sep = "vs")
signif.results <- sta.results %>% 
  filter(p_value < 0.05) %>% 
  filter(adjust_method == "Bonferroni")
signif.results.tukey <- sta.results.tukey %>% 
  filter(p_value < 0.05)
```

**Diversity beta**

```{r message=FALSE, warning=FALSE}
# Dataframe
permanovaDF <- data.frame(Obj.Phyloseq = character(),
                          Distance = character(),
                          F_value = numeric(),
                          p_value = numeric(),
                          Versus = character(),
                          p_value_pairs = numeric())
for (o.p in obj.phylo){
  # o.p <- "Genus_RR"
  ob.phy <- get(o.p)
  Wunifrac <- UniFrac(physeq = ob.phy, weighted = TRUE, normalized = TRUE, 
                      parallel = TRUE, fast = TRUE)
  permanova <- adonis2(Wunifrac ~ sample_data(ob.phy)$SITE)
  permanova.pair <- pairwise.adonis(Wunifrac, sample_data(ob.phy)$SITE)
  permanovaDF <- bind_rows(permanovaDF, 
                           data.frame(Obj.Phyloseq = rep(o.p, 3),
                                      Distance = rep("Weighted UniFrac", 3),
                                      F_value = rep(permanova$F[1], 3),
                                      p_value = rep(permanova$`Pr(>F)`[1], 3),
                                      Versus = permanova.pair$pairs,
                                      p_value_pairs = permanova.pair$p.value))
}
signif.permanova <- permanovaDF %>% 
  filter(p_value_pairs < 0.05)
signif.permanova <- signif.permanova[order(signif.permanova$Obj.Phyloseq),]
signif.permanova
```

## Different visualizations of abundances

**Object Ampvis2**

```{r, message = FALSE, warning = FALSE}
obj.phylo <- c("Phylum_RRR", "Genus_RRR")

# Cycle for each object phyloseq
for (i in obj.phylo) {
  # i <- "Genus_RRR"
  phylo_obj <- get(i)
  otu_table(phylo_obj) <- t(otu_table(phylo_obj))
  Otu_df <- data.frame(OTU = rownames(phyloseq::otu_table(phylo_obj)@.Data),
                       phyloseq::otu_table(phylo_obj)@.Data,
                       phyloseq::tax_table(phylo_obj)@.Data,
                       check.names = FALSE)
  Otu_df$Species <- Otu_df$Genus
  Met_df <- data.frame(phyloseq::sample_data(phylo_obj), 
                       check.names = FALSE)
  Met_df$SAMPLE <- rownames(Met_df)
  Met_df <- Met_df[,c(10, 1:9)]
  Amp <- amp_load(Otu_df, Met_df)
  assign(paste0("AV_", i), Amp, envir = .GlobalEnv)
}
```

**Heat map**

```{r}
av <- "AV_Phylum_RRR"
avobj <- get(av)
av.plot <- amp_heatmap(avobj, group_by = "SITE", facet_by = "SITE", 
                       plot_values = TRUE, tax_show = 10, 
                       normalise = FALSE,
                       tax_aggregate = "Phylum", 
                       color_vector = brewer.pal(9, "BrBG"),
                       plot_colorscale = "sqrt") +
  labs(title = paste(av, "abundance for Phylum"))
av.plot

av <- "AV_Genus_RRR"
avobj <- get(av)
av.plot <- amp_heatmap(avobj, group_by = "SITE", facet_by = "SITE", 
                       plot_values = TRUE, tax_show = 10, 
                       normalise = FALSE, 
                       tax_aggregate = "Phylum",
                       color_vector = brewer.pal(9, "BrBG"),
                       plot_colorscale = "sqrt") +
  labs(title = paste(av, "abundance for Phylum"))
av.plot2 <- amp_heatmap(avobj, group_by = "SITE", facet_by = "SITE", 
                       plot_values = TRUE, tax_show = 10, 
                       normalise = FALSE,
                       tax_aggregate = "Genus", 
                       tax_add = "Phylum",
                       color_vector = brewer.pal(9, "BrBG"),
                       plot_colorscale = "sqrt") +
  labs(title = paste(av, "abundance for Phylum and Genus"))
```

**Boxplot**

```{r, message = FALSE, warning = FALSE}
# Top
ntop <- 10
i <- "Phylum_RRR_DF"
df <- get(i)
tax <- df %>% 
  group_by(Phylum) %>% 
  dplyr::summarise(TotalAbundance = sum(Abundance))
top <- tax %>% 
  arrange(dplyr::desc(TotalAbundance)) %>% 
  head(ntop)
df_filter <- df %>% 
  filter(Phylum %in% top$Phylum)
# Plot
plot <- ggplot(df_filter, aes(x = SITE, y = Abundance, fill = SITE)) +
  geom_boxplot() +
  # Scale
  labs(title = paste("Abundance at the phylum taxonomic level with", i), 
       x = "Site", y = "Abundance", fill = tx) +
  facet_grid(YEAR ~ Phylum)

i <- "Genus_RRR_DF"
df <- get(i)
for (j in c("Phylum", "Genus")){
  # j <- "Phylum"
  tax <- df %>% 
    group_by(get(j)) %>% 
    dplyr::summarise(TotalAbundance = sum(Abundance))
  top <- tax %>% 
    arrange(dplyr::desc(TotalAbundance)) %>% 
    head(ntop)
  df_filter <- df %>% 
    filter(get(j) %in% top$`get(j)`)
  # Plot
  plot <- ggplot(df_filter, aes(x = SITE, y = Abundance, fill = SITE)) +
    geom_boxplot() +
    # Scale
    labs(title = paste("Abundance at the", j, "taxonomic level with", i), 
         x = "Site", y = "Abundance", fill = tx) +
    facet_grid(YEAR ~ get(j))
}
```

**Pheatmap**

```{r message=FALSE, warning=FALSE}
# Colors
color.met <- list(SITE = c("Ag-NP1" = "#004586", 
                           "Ag-NP2" = "#579D1C", 
                           "Ag-SU3" = "#C5000B"),
                  YEAR = c("2017" = "#FFD320", 
                           "2018" = "#4B1F6F", 
                           "2019" = "#DD4477"))
color.hm <- brewer.pal(10, "Blues")

# Top
ntop <- 10

o.p <- "Phylum_RRR"
obj.p <- get(o.p)
Otu <- merge_phyloseq(otu_table(t(obj.p@otu_table@.Data), taxa_are_rows = T), 
                      obj.p@tax_table, obj.p@sam_data, obj.p@phy_tree, obj.p@refseq)
Otu <- tax_glom(Otu, taxrank = rank_names(Otu)[2]) 
Top <- sort(taxa_sums(Otu), TRUE)[1:ntop]
Top <- prune_taxa(names(Top), Otu)
Otu_df <- as.data.frame(Top@otu_table@.Data)
rownames(Otu_df) <- make.names(Top@tax_table@.Data[,2], unique = T)
Met_df <- data.frame(sample_data(Top)[,c(1,5)])
plot.hm <- pheatmap(main = paste(o.p, "values at the taxonomic level of Phylum"),
                    Otu_df, color = color.hm, cluster_cols = TRUE, 
                    border_color = "#000000", 
                    annotation_col = Met_df, annotation_colors = color.met)
plot.hm

o.p <- "Genus_A"
# phylum
obj.p <- get(o.p)
Otu <- merge_phyloseq(otu_table(t(obj.p@otu_table@.Data), taxa_are_rows = T), 
                      obj.p@tax_table, obj.p@sam_data, obj.p@phy_tree, obj.p@refseq)
Otu <- tax_glom(Otu, taxrank = rank_names(Otu)[2]) 
Top <- sort(taxa_sums(Otu), TRUE)[1:ntop]
Top <- prune_taxa(names(Top), Otu)
Otu_df <- as.data.frame(Top@otu_table@.Data)
rownames(Otu_df) <- make.names(Top@tax_table@.Data[,2], unique = T)
Met_df <- data.frame(sample_data(Top)[,c(1,5)])
plot.hm <- pheatmap(main = paste(o.p, "values at the taxonomic level of Phylum"),
                    Otu_df, color = color.hm, cluster_cols = TRUE, 
                    border_color = "#000000", 
                    annotation_col = Met_df, annotation_colors = color.met)
plot.hm
# Genus
Otu <- merge_phyloseq(otu_table(t(obj.p@otu_table@.Data), taxa_are_rows = T), 
                      obj.p@tax_table, obj.p@sam_data, obj.p@phy_tree, obj.p@refseq)
Otu <- tax_glom(Otu, taxrank = rank_names(Otu)[6]) 
Top <- sort(taxa_sums(Otu), TRUE)[1:ntop]
Top <- prune_taxa(names(Top), Otu)
Otu_df <- as.data.frame(Top@otu_table@.Data)
rownames(Otu_df) <- make.names(Top@tax_table@.Data[,6], unique = T)
Met_df <- data.frame(sample_data(Top)[,c(1,5)])
plot.hm2 <- pheatmap(main = paste(o.p, "values at the taxonomic level of Genus"),
                    Otu_df, color = color.hm, cluster_cols = TRUE, 
                    border_color = "#000000",
                    annotation_col = Met_df, annotation_colors = color.met)
plot.hm2
```

# Scripts Shotgun
## Import and manipulation of kraken-biom file and metadata

**Import**

```{r}
merged_metagenomes <- import_biom(paste0(Shotgun, "Biom/CalakmulShotGun.biom"))
class(merged_metagenomes)
head(merged_metagenomes@tax_table@.Data)
```

**Manipulation**

```{r}
merged_metagenomes@tax_table@.Data <- substring(merged_metagenomes@tax_table@.Data, 4)
colnames(merged_metagenomes@tax_table@.Data) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
head(merged_metagenomes@tax_table@.Data)
TablaMergedMetagenomes <- merged_metagenomes@tax_table@.Data
write.table(TablaMergedMetagenomes, 
            file = paste0(Shotgun, "CalakmulTablaMergedMetagenomes.txt"), 
            row.names=TRUE, col.names=NA, quote=FALSE, sep="\t")
```

**Filter**

```{r}
merged_metagenomes <- subset_taxa(merged_metagenomes, Kingdom == "Bacteria")
```

**Metadata**

```{r}
MetaShotgun <- data.frame(SITE = c("Ag1", "Ag2", "Ag3"), 
                          pH = c(6.2, 6.7, 6.1), 
                          N_TOTAL = c(0.36, 0.22, 0.29),
                          P_OLSEN = c(29.6, 13.5, 4.0),
                          ORGANIC = c(12.0, 7.8, 24.2))
rownames(MetaShotgun) <- MetaShotgun$SITE
rownames(merged_metagenomes) %in% rownames(MetaShotgun)
all(rownames(merged_metagenomes) %in% MetaShotgun$SITE)
merged_metagenomes@sam_data <- sample_data(MetaShotgun)
```

## Data transformation and manipulation

**Prune and subset taxa**

```{r, message = FALSE, warning = FALSE}
# How much empty data there is 
summary(merged_metagenomes@tax_table@.Data== "") 
ShotGun <- prune_taxa(taxa_sums(merged_metagenomes) > 0, merged_metagenomes)
Phylo.sg_P <- subset_taxa(ShotGun, Phylum != "")
Phylo.sg_G <- subset_taxa(ShotGun, Genus != "")
summary(Phylo.sg_P@tax_table@.Data == "")  
summary(Phylo.sg_G@tax_table@.Data == "")  
```

**Rarefacction data**

```{r message=FALSE, warning=FALSE}
# Grouping by taxonomic level
Phylum.sg_A <- tax_glom(physeq = Phylo.sg_P, taxrank = "Phylum", NArm = F)
Genus.sg_A <- tax_glom(physeq = Phylo.sg_G, taxrank = "Genus", NArm = F)
# Rarefaction Species Richness
Phylum.sg_RR <- rarefy_even_depth(Phylum.sg_A, sample.size = min(sample_sums(Phylum.sg_A)),
                               rngseed = 711, replace = TRUE, 
                               trimOTUs = TRUE, verbose = TRUE)
Genus.sg_RR <- rarefy_even_depth(Genus.sg_A, sample.size = min(sample_sums(Genus.sg_A)),
                              rngseed = 711, replace = TRUE, 
                              trimOTUs = TRUE, verbose = TRUE)
# Rarefaction relative data
Phylum.sg_RRR <- transform_sample_counts(Phylum.sg_RR, function(x) round(x*100 / sum(x), 0))
Genus.sg_RRR <- transform_sample_counts(Genus.sg_RR, function(x) round(x*100 / sum(x), 0))
```

From here on, in the sections on Data transformation and manipulation, Abundance plots, Diversity analysis, and Different abundance visualizations, the script is similar to what was done with rRNA 16s, taking as input the objects: Phylum.sg_RR, Genus.sg_RR, Phylum.sg_RRR, Genus.sg_RRR, Genus.sg_RRR, as the case may be.

## Virus

For virus-related metagenomics analysis, the following tools were used: VIBRANT (https://github.com/AnantharamanLab/VIBRANT) and Kaiju (https://github.com/bioinformatics-centre/kaiju).

# Session information

```{r}
sessionInfo()
```
